examples/llm/src/models/layers/gpt_blocks.py

```diff
@@ -47,20 +52,64 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
 
     def forward(
         self,
+        a: torch.Tensor,
         x: torch.Tensor,
         past_key_value: Optional[Tuple[torch.Tensor]] = None,
         attn_bias: Optional[torch.Tensor] = None,
         key_padding_mask: Optional[torch.ByteTensor] = None,
         is_causal: bool = True,
-    ) -> Tuple[torch.Tensor, Optional[Tuple[torch.Tensor]]]:
-        a = self.ln_1(x)
+    ) -> Tuple[torch.Tensor, torch.Tensor, Optional[Tuple[torch.Tensor]]]:
         b, _, past_key_value = self.attn(a,
                                          past_key_value=past_key_value,
                                          attn_bias=attn_bias,
                                          key_padding_mask=key_padding_mask,
                                          is_causal=is_causal)
         x = x + self.resid_attn_dropout(b)
-        m = self.ln_2(x)
+        m = self.ln_1(x)
         n = self.mlp(m)
         x = x + self.resid_mlp_dropout(n)
-        return x, past_key_value
+        a = self.ln_2(x)
+        return a, x, past_key_value
+
+
+class OptimizedGPTBlock(nn.Module):
+
+    def __init__(self, cfg: DictConfig, device: Optional[str] = None):
+        super().__init__()
+
+        if not DROPOUT_ADD_LAYERNORM_INSTALLED:
+            raise ImportError(
+                'A CUDA optimization required for the Optimized GPT Block were not installed. Please install it from the examples/llm/csrc folder.'
+            )
+
+        self.attn = MultiheadAttention(cfg, device)
+        self.dropout_add_ln_1 = DropoutAddLayerNorm(cfg.d_model,
+                                                    prenorm=True,
```

Comment:
What does `prenorm=True` mean?
Does this mean make it a pre(layer)norm block?

Is GPTMLP and OptimizedGPTBlock exactly the same?
If yes, can we add testes to verify this? (along with bwd pass tests?)